### PR TITLE
Living settlor - content and form validation

### PR DESCRIPTION
**Pull request by JoshuaH01**

### CODE REVIEW

File: app/forms/PassportOrIdCardFormProvider.scala [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+            isNotEmpty("number", "settlorIndividualPassport.number.error.required")
+          )
+        ),
+      "expiryDate" -> localDate(
+        invalidKey     = "settlorIndividualPassport.expiryDate.error.invalid",
+        allRequiredKey = "settlorIndividualPassport.expiryDate.error.required.all",
+        twoRequiredKey = "settlorIndividualPassport.expiryDate.error.required.two",
+        requiredKey    = "settlorIndividualPassport.expiryDate.error.required"
+      ).verifying(firstError(
+        maxDate(LocalDate.now, s"settlorIndividualPassport.expiryDate.error.future", "day", "month", "year"),
```
Comment (jassalrichy) [inline app/forms/PassportOrIdCardFormProvider.scala]: ID Card can have a future expiry date, this only allows up to today.
Error messages should relate to ID Card and not Passport.

File: test/views/SettlorIndividualOrBusinessViewSpec.scala [hunk 2]
```diff
@@ -17,13 +17,13 @@
...
-import models.{NormalMode, IndividualOrBusiness}
+import models.{IndividualOrBusiness, NormalMode}
 import play.api.data.Form
 import play.twirl.api.HtmlFormat
-import views.behaviours.ViewBehaviours
+import views.behaviours.{OptionsViewBehaviours, ViewBehaviours}
 import views.html.SettlorIndividualOrBusinessView
 
-class SettlorIndividualOrBusinessViewSpec extends ViewBehaviours {
+class SettlorIndividualOrBusinessViewSpec extends OptionsViewBehaviours {
```
Comment (adamconder) [inline test/views/SettlorIndividualOrBusinessViewSpec.scala]: Is this changing the existing implementation for the DeceasedSettlorIndividual Or Business question?

File: app/forms/PassportOrIdCardFormProvider.scala (hunk 1 above)
Comment (adamconder) [reply app/forms/PassportOrIdCardFormProvider.scala]: This has now been set to 2100 as the maximum expiry date.

File: app/forms/SettlorIndividualNINOFormProvider.scala [hunk 3]
```diff
@@ -24,8 +24,12 @@ import play.api.data.Form
...
 
   def apply(): Form[String] =
-    Form(
-      "value" -> text("settlorIndividualNINO.error.required")
-        .verifying(maxLength(100, "settlorIndividualNINO.error.length"))
-    )
+    Form("value" -> nino("settlorIndividualNINO.error.required")
+      .verifying(
+        firstError(
+          isNotEmpty("value", "settlorIndividualNINO.error.required"),
```
[the remark this answers is not in the document]
Comment (adamconder) [reply app/forms/SettlorIndividualNINOFormProvider.scala]: Implemented a generic NinoFormProvider and removed the duplicate SettlorNinoForm.

File: app/forms/Validation.scala [hunk 4]
```diff
@@ -33,7 +33,7 @@ object Validation {
   val decimalCheck = "^[^.]*$"
   val validNinoFormat: String = "[[A-Z]&&[^DFIQUV]][[A-Z]&&[^DFIQUVO]] ?\\d{2} ?\\d{2} ?\\d{2} ?[A-D]{1}"
   val descriptionRegex = "^[0-9a-zA-Z{\\u00C0-\\u02FF\\u2019} \\u005C&`'^\\-]*$"
-
+  val passportOrIdCardNumberRegEx = """^([A-Za-z0-9]{1,30})$"""
```
[the remark this answers is not in the document]
Comment (jassalrichy) [reply app/forms/Validation.scala]: pattern confirmed as per des spec 5.0

File: conf/messages.en [hunk 5]
```diff
@@ -1053,89 +1051,111 @@ settlorIndividualName.error.firstname.invalid = Settlor’s first name must only
...
-settlorIndividualNINOYesNo.heading = settlorIndividualNINOYesNo
-settlorIndividualNINOYesNo.checkYourAnswersLabel = settlorIndividualNINOYesNo
-settlorIndividualNINOYesNo.error.required = Select yes if settlorIndividualNINOYesNo
-
-settlorIndividualAddressInternational.title = settlorIndividualAddressInternational
-settlorIndividualAddressInternational.heading = settlorIndividualAddressInternational
+settlorIndividualDateOfBirthYesNo.title = Do you know the settlor’s date of birth?
+settlorIndividualDateOfBirthYesNo.heading = Do you know {0}’s date of birth?
+settlorIndividualDateOfBirthYesNo.checkYourAnswersLabel = Do you know {0}’s date of birth?
+settlorIndividualDateOfBirthYesNo.error.required = Select yes if you know date of birth
```
[the remark this answers is not in the document]
Comment (jassalrichy) [reply conf/messages.en]: TODO

File: conf/messages.en [hunk 6]
```diff
@@ -1053,89 +1051,111 @@ settlorIndividualName.error.firstname.invalid = Settlor’s first name must only
...
+
+settlorIndividualDateOfBirth.title = What is the settlor’s date of birth?
+settlorIndividualDateOfBirth.heading = What is {0}’s date of birth?
+settlorIndividualDateOfBirth.checkYourAnswersLabel = What is {0}’s date of birth?
+settlorIndividualDateOfBirth.hint = For example, 31 3 1980
+settlorIndividualDateOfBirth.error.required.all = Enter a date of birth
+settlorIndividualDateOfBirth.error.required.two = The date of birth must include {0} and {1}
```
[the remark this answers is not in the document]
Comment (jassalrichy) [reply conf/messages.en]: TODO, as part of discussions with Anita.

File: conf/messages.en [hunk 7]
```diff
@@ -1053,89 +1051,111 @@ settlorIndividualName.error.firstname.invalid = Settlor’s first name must only
...
+settlorIndividualIDCard.checkYourAnswersLabel = What are {0}’s ID card details?
+settlorIndividualIDCard.country = ID card country of issue
+settlorIndividualIDCard.error.country.required = Enter the settlor’s ID card country of issue
+settlorIndividualIDCard.number = ID card number
+settlorIndividualIDCard.number.error.required = Enter the settlor’s ID card number
+settlorIndividualIDCard.number.error.length = Passport or ID card number must be 30 characters or less
+settlorIndividualIDCard.number.error.invalid = Passport or ID card number can only include letters and numbers
+settlorIndividualIDCard.expiryDate.title = Expiry date
+settlorIndividualIDCard.expiryDate.hint = Expiry date can be in the past. For example, 31 3 1980
+settlorIndividualIDCard.error.required.all = Enter an expiry date
```
[the remark this answers is not in the document]
Comment (jassalrichy) [reply conf/messages.en]: TODO, as per conversations with Anita.

File: test/forms/behaviours/DateBehaviours.scala [hunk 8]
```diff
@@ -44,6 +45,29 @@ class DateBehaviours extends FieldBehaviours {
...
+  def dateFieldForPassportOrIdForm(form: Form[PassportOrIdCardDetails],
+                                   key: String, validData: Gen[LocalDate],
+                                   requiredBindings: Map[String,String] = Map.empty): Unit = {
+
+    "bind valid data" in {
+
+      forAll(validData -> "valid date") {
+        date =>
+
+          val data = requiredBindings ++ Map(
```
[the remark this answers is not in the document]
Comment (jassalrichy) [reply test/forms/behaviours/DateBehaviours.scala]: Need to confirm if this is causing an issue?

File: test/views/SettlorIndividualOrBusinessViewSpec.scala (hunk 2 above)
Comment (jassalrichy) [reply test/views/SettlorIndividualOrBusinessViewSpec.scala]: This is a new implementation and not part of the deceased settlor journey

File: app/controllers/SettlorIndividualAddressInternationalController.scala [hunk 9]
```diff
@@ -50,25 +54,32 @@ class SettlorIndividualAddressInternationalController @Inject()(
...
+      Ok(view(preparedForm, countryOptions.options, mode, index, draftId, name))
   }
 
   def onSubmit(mode: Mode, index: Int, draftId: String): Action[AnyContent] = actions(index, draftId).async {
     implicit request =>
 
+      val name = request.userAnswers.get(SettlorIndividualNamePage(index)).get
+
       form.bindFromRequest().fold(
         (formWithErrors: Form[_]) =>
```
[the remark this answers is not in the document]
Comment (jassalrichy) [reply app/controllers/SettlorIndividualAddressInternationalController.scala]: Resolved.

File: app/pages/SettlorIndividualAddressInternationalPage.scala [hunk 10]
```diff
@@ -18,10 +18,11 @@ package pages
...
 import play.api.libs.json.JsPath
+import sections.Settlors
 
 final case class SettlorIndividualAddressInternationalPage(index : Int) extends QuestionPage[InternationalAddress] {
 
-  override def path: JsPath = JsPath \ toString
+  override def path: JsPath = JsPath \ Settlors \ "living" \ toString
 
-  override def toString: String = "settlorIndividualAddressInternational"
+  override def toString: String = "internationalAddress"
```
[the remark this answers is not in the document]
Comment (jassalrichy) [reply app/pages/SettlorIndividualAddressInternationalPage.scala]: Done.

File: app/pages/SettlorIndividualAddressUKPage.scala [hunk 11]
```diff
@@ -18,10 +18,11 @@ package pages
...
 import play.api.libs.json.JsPath
+import sections.Settlors
 
 final case class SettlorIndividualAddressUKPage(index : Int) extends QuestionPage[UKAddress] {
 
-  override def path: JsPath = JsPath \ toString
+  override def path: JsPath = JsPath \ Settlors \ "living"  \ toString
 
-  override def toString: String = "settlorIndividualAddressUK"
+  override def toString: String = "ukAddress"
```
[the remark this answers is not in the document]
Comment (jassalrichy) [reply app/pages/SettlorIndividualAddressUKPage.scala]: Done.

File: app/pages/SettlorIndividualAddressInternationalPage.scala [hunk 12]
```diff
@@ -18,10 +18,11 @@ package pages
 
 import models.InternationalAddress
 import play.api.libs.json.JsPath
+import sections.Settlors
 
 final case class SettlorIndividualAddressInternationalPage(index : Int) extends QuestionPage[InternationalAddress] {
 
-  override def path: JsPath = JsPath \ toString
+  override def path: JsPath = JsPath \ Settlors \ "living" \ toString
```
[the remark this answers is not in the document]
Comment (jassalrichy) [reply app/pages/SettlorIndividualAddressInternationalPage.scala]: Resolved.

File: app/utils/CheckYourAnswersHelper.scala [hunk 13]
```diff
@@ -46,9 +46,9 @@ class CheckYourAnswersHelper @Inject()(countryOptions: CountryOptions)(userAnswe
   def settlorIndividualPassport(index: Int): Option[AnswerRow] = userAnswers.get(SettlorIndividualPassportPage(index)) map {
     x =>
       AnswerRow(
-        "settlorIndividualPassport.checkYourAnswersLabel",
-        HtmlFormat.escape(s"${x.field1} ${x.field2}"),
-        routes.SettlorIndividualPassportController.onPageLoad(CheckMode, index,  draftId).url
+        "individualBeneficiariesIDCardDetails.checkYourAnswersLabel",
```
[the remark this answers is not in the document]
Comment (jassalrichy) [reply app/utils/CheckYourAnswersHelper.scala]: Corrected.

File: app/utils/CheckYourAnswersHelper.scala [hunk 14]
```diff
@@ -46,9 +46,9 @@ class CheckYourAnswersHelper @Inject()(countryOptions: CountryOptions)(userAnswe
...
+        HtmlFormat.escape(s"${x.country} ${x.cardNumber}, ${x.expiryDate}"),
```
[the remark this answers is not in the document]
Comment (jassalrichy) [reply app/utils/CheckYourAnswersHelper.scala]: Corrected.

File: app/utils/CheckYourAnswersHelper.scala [hunk 15]
```diff
@@ -64,9 +64,9 @@ class CheckYourAnswersHelper @Inject()(countryOptions: CountryOptions)(userAnswe
   def settlorIndividualIDCard(index: Int): Option[AnswerRow] = userAnswers.get(SettlorIndividualIDCardPage(index)) map {
     x =>
       AnswerRow(
-        "settlorIndividualIDCard.checkYourAnswersLabel",
-        HtmlFormat.escape(s"${x.field1} ${x.field2}"),
-        routes.SettlorIndividualIDCardController.onPageLoad(CheckMode, index,  draftId).url
+        "individualBeneficiariesIDCardDetails.checkYourAnswersLabel",
```
[the remark this answers is not in the document]
Comment (jassalrichy) [reply app/utils/CheckYourAnswersHelper.scala]: Corrected.

File: app/views/SettlorIndividualIDCardView.scala.html [hunk 16]
```diff
@@ -34,16 +36,28 @@
...
         @components.error_summary(form.errors)
 
-        @components.heading("settlorIndividualIDCard.heading")
+        @components.heading("settlorIndividualIDCard.heading", Some(name.toString))
 
-        @components.input_text(
-            field = form("field1"),
-            label = messages("settlorIndividualIDCard.field1")
+        @components.select(field = form("country"),
+            label = messages("settlorIndividualIDCard.country"),
```
[the remark this answers is not in the document]
Comment (jassalrichy) [reply app/views/SettlorIndividualIDCardView.scala.html]: TODO.Need to discuss with Anita

File: app/views/SettlorIndividualIDCardView.scala.html [hunk 17]
```diff
@@ -34,16 +36,28 @@
...
+            placeholder = Some("address.country.select"),
+            formGroupClasses = Set("autocomplete-wrapper"),
+            selectAttribute = Set("data-non-uk-countries")
         )
 
         @components.input_text(
-            field = form("field2"),
-            label = messages("settlorIndividualIDCard.field2")
+            field = form("number"),
+            label = messages("settlorIndividualIDCard.number")
```
[the remark this answers is not in the document]
Comment (jassalrichy) [reply app/views/SettlorIndividualIDCardView.scala.html]: TODO. Need to discuss with Anita.

File: app/views/SettlorIndividualPassportView.scala.html [hunk 18]
```diff
@@ -34,16 +36,29 @@
...
         @components.error_summary(form.errors)
 
-        @components.heading("settlorIndividualPassport.heading")
+        @components.heading("settlorIndividualPassport.heading", Some(name.toString))
 
-        @components.input_text(
-            field = form("field1"),
-            label = messages("settlorIndividualPassport.field1")
+        @components.select(field = form("country"),
+            label = messages("settlorIndividualPassport.country"),
```
[the remark this answers is not in the document]
Comment (jassalrichy) [reply app/views/SettlorIndividualPassportView.scala.html]: TODO. Need to discuss with Anita.

File: app/views/SettlorIndividualPassportView.scala.html [hunk 19]
```diff
@@ -34,16 +36,29 @@
...
+            placeholder = Some("address.country.select"),
+            formGroupClasses = Set("autocomplete-wrapper"),
+            selectAttribute = Set("data-non-uk-countries")
         )
 
         @components.input_text(
-            field = form("field2"),
-            label = messages("settlorIndividualPassport.field2")
+            field = form("number"),
+            label = messages("settlorIndividualPassport.number"),
```
[the remark this answers is not in the document]
Comment (jassalrichy) [reply app/views/SettlorIndividualPassportView.scala.html]: TODO. Need to discuss with Anita.

File: app/views/SettlorIndividualPassportView.scala.html [hunk 20]
```diff
@@ -34,16 +36,29 @@
...
+            hint = Some(messages(s"settlorIndividualPassport.number.hint"))
```
[the remark this answers is not in the document]
Comment (jassalrichy) [reply app/views/SettlorIndividualPassportView.scala.html]: TODO. Need to discuss with Anita.

File: test/controllers/SettlorIndividualDateOfBirthControllerSpec.scala [hunk 21]
```diff
@@ -165,5 +186,6 @@ class SettlorIndividualDateOfBirthControllerSpec extends SpecBase with MockitoSu
 
       application.stop()
     }
+
```
[the remark this answers is not in the document]
Comment (jassalrichy) [reply test/controllers/SettlorIndividualDateOfBirthControllerSpec.scala]: Added.

File: test/controllers/SettlorIndividualNameControllerSpec.scala [hunk 22]
```diff
@@ -149,5 +150,38 @@ class SettlorIndividualNameControllerSpec extends SpecBase {
 
       application.stop()
     }
+
```
[the remark this answers is not in the document]
Comment (jassalrichy) [reply test/controllers/SettlorIndividualNameControllerSpec.scala]: Amended.

File: test/forms/PassportOrIdCardFormProviderSpec.scala [hunk 23]
```diff
@@ -0,0 +1,133 @@
...
+
+class PassportOrIdCardFormProviderSpec extends
+  StringFieldBehaviours with PassportOrIDCardBehaviours with DateBehaviours {
+
+  val form = new PassportOrIdCardFormProvider()()
+
+  ".country" must {
+
+    val fieldName = "country"
+    val requiredKey = "settlorIndividualPassport.country.error.required"
```
[the remark this answers is not in the document]
Comment (jassalrichy) [reply test/forms/PassportOrIdCardFormProviderSpec.scala]: Resolved.

File: test/forms/SettlorIndividualDateOfBirthFormProviderSpec.scala [hunk 24]
```diff
@@ -25,15 +25,29 @@ class SettlorIndividualDateOfBirthFormProviderSpec extends DateBehaviours {
...
+      max = max
     )
 
     behave like dateField(form, "value", validData)
 
     behave like mandatoryDateField(form, "value", "settlorIndividualDateOfBirth.error.required.all")
+
+    behave like dateFieldWithMax(form, "value",
+      max = max,
+      FormError("value", s"settlorIndividualDateOfBirth.error.future", List("day", "month", "year"))
```
[the remark this answers is not in the document]
Comment (jassalrichy) [reply test/forms/SettlorIndividualDateOfBirthFormProviderSpec.scala]: Generic DateOfBirthFormProvider created.

File: app/controllers/SettlorIndividualAddressInternationalController.scala [hunk 25]
```diff
@@ -50,25 +54,32 @@ class SettlorIndividualAddressInternationalController @Inject()(
   private def actions(index: Int, draftId: String) =
     identify andThen
       getData(draftId) andThen
-      requireData
+      requireData andThen
+      validateIndex(index, Settlors) andThen
```
[the remark this answers is not in the document]
Comment (jassalrichy) [reply app/controllers/SettlorIndividualAddressInternationalController.scala]: Applied.

**Pull request merged**